### ISSUE TITLE
bug(cuda): --backend cuda serve emits degenerate '!' output (device-GEMM accuracy) — live on L4 sm_89 + A100 sm_80, CPU path correct

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#3186]
## Symptom

`fak serve --backend cuda` emits **deterministic degenerate output** — an unbroken run of the single token `!` from generation token 0 — for a small dense model, while the **byte-identical** GGUF served on fak's CPU backend decodes correctly.

Witnessed live 2026-07-08 on GCP `fak-realmodel` (g2-standard-8, **NVIDIA L4, sm_89**, driver 535.309.01), model `qwen2.5-0.5b-instruct-q8_0.gguf`, temp=0:

| endpoint | backend | output |
|---|---|---|
| `:8082` | `fak serve --backend cuda` | `'!!!!!!!!!!!!...'` (200/200 tokens, finish=length) |
| `:8081` | `fak serve` (CPU, **same gguf**) | `'Paris'` / `'Red'` ✅ |
| `:8080` | ollama qwen2.5:14b (reference) | `'Paris'` ✅ |

3/3 deterministic. Isolation is clean: same prompt, same model file, same gateway — the only variable is `--backend cuda`.

## Root cause (characterized)

- **Device-GEMM numerical-accuracy bug**, not a NaN/collapse. A temperature sweep (temp 0 / 0.7 / 1.5 / 2.0) yields **wrong-but-STRUCTURED, deterministic** logits (greedy → all-`!`; sampled → varied multilingual garbage), which rules out NaN/inf/constant-collapse and points at wrong GEMM math. Degenerate-from-token-0 pins it to the forward/GEMM kernels, not KV-cache or sampling.
- **GPU utilization stays 0% at idle power (~33 W)** during a 200-token CUDA decode with weights resident (2254 MiB) — the decode path does no observable GPU compute yet still emits output (possibly a broken CUDA-graph decode replay; cf. #3076 L4 flash-attn/CUDA-graph work).

## Cross-arch: not L4-specific

Corroborates + extends the ledgered DGX **A100 sm_80** device-GEMM accuracy failures (`TestCUDAQ8BatchedMatMulApproxMatchesRef`, `TestCUDAQ4KMatMulApproxMatchesRef` FAIL; degenerate GLM-5.2 Q4_K device output) from 2026-06-30/07-01 to a **2nd GPU arch (sm_89)** and a 2nd model/quant via a live serve. Related to #972 (CUDA MatMul wrong output length for non-block-aligned dims, sm_80) — same subsystem, different manifestation.

## Evidence (nightrun ledger)

`docs/nightrun/collected.jsonl` — task_ids `witness-cuda-backend-degenerate-output-l4` (commit `09318126`, dos commit-audit OK/diff-witnessed) + temp-sweep rows; DGX A100 rows dated 2026-06-30/07-01.

## Secondary finding: no serve output-coherence gate

The `:8082` cuda serve had been UP ~12 days silently returning this garbage at HTTP 200 / 0.027 s — `fak serve`'s `/healthz` is liveness-only; no capability asserts the loaded backend decodes a non-degenerate first token. A status-code/latency monitor reads a garbage-producing backend as healthy.

## Next checkable steps

1. Bind the live-serve symptom to the kernel test: on an sm_89 (and sm_80) **Linux+CUDA+Go build host**, run `go test -tags cuda -run TestCUDAQ8BatchedMatMulApproxMatchesRef`. (Currently blocked for lack of a build host — the GCP serve boxes have the GPU + a cuda-capable fak binary but no Go/nvcc.)
2. Add a **fail-closed serve readiness probe**: fixed-prompt smoke on load → fail `/healthz` (or gate a `/ready`) when the first decoded token is a repeated degenerate id, so a broken backend cannot advertise healthy.

Filed from a live GCP L4 testing pass; honest gap: the box's on-disk `/usr/local/bin/fak` is unlinked (serves run from a deleted inode) so the exact serving SHA could not be read on-box.

**Comment by anthony-chaudhary** [comment]
**Reconciliation — next-step #1 (bind serve symptom → kernel unit test) is DONE, not blocked.**

Same-day this was witnessed at the unit level on a dedicated GCP L4 build host:

- host `fak-cuda-build-l4` (g2-standard-8, **NVIDIA L4 sm_89**, us-central1-b) — spun up purpose=cuda-unit-repro
- nvcc-built `libfakcuda.a` for sm_89 → 240538-byte archive, `go build -tags cuda ./internal/compute` OK (`BUILD_RC=0`)
- `go test -tags cuda ./internal/compute ./internal/model` → **`--- FAIL: TestCUDAForwardMatchesRef`** (`FAIL … /internal/compute 1.743s`), `ok … /internal/model`

So two independent witnesses (live serve + unit test) now agree the fak CUDA backend is broken on **L4/sm_89**, and this extends the ledgered DGX A100/sm_80 `TestCUDAForwardMatchesRef` failures (2026-06-30) to sm_89. Evidence: `docs/nightrun/collected.jsonl` task_id `witness-cuda-forwardmatchesref-unit-fail-l4-sm89`, commit `872877c31`.

**Honest gaps carried forward (this is step 1b, not fully closed):**
1. build ran from an ephemeral clone since cleaned → exact serving/build SHA not pinned;
2. only the default-verbosity FAIL line captured — **not** the per-test want/got argmax diff that would name the failing GEMM op.

**Resource state (so the next session doesn't re-provision):** `fak-cuda-build-l4` was reaped to **TERMINATED** when idle (`fak-reaper=reap`), but its **150GB pd-balanced disk is retained** — it holds the working sm_89 nvcc/CUDA/Go build env. Restart it (don't recreate) to do the remaining step-1b: re-clone at a **pinned** `origin/main` SHA, rebuild `-tags cuda`, capture `go test -v -run TestCUDAForwardMatchesRef` want/got to name the exact op, then bisect against the `ce93da02` uploadCache-guard lineage.

_Note: these commits are in the shared local trunk but not yet on `origin/main` — this host's HTTPS git-push egress is currently down (reads fast, writes hang); they'll fast-forward once egress recovers._

**Comment by anthony-chaudhary** [comment]
**Correction + the exact failing op (verbose `-tags cuda` run on L4/sm_89).**

⚠️ My previous comment (and the ledger row `witness-cuda-forwardmatchesref-unit-fail-l4-sm89`, commit `872877c31`) said `TestCUDAForwardMatchesRef` FAILs on sm_89. **That was wrong** — it was a **build/link failure misread as an accuracy fail**: `go test -tags cuda` couldn't find `-lcudart`/`-lcublas` without `LIBRARY_PATH=/usr/local/cuda/lib64`, yielding a bare `FAIL … [build failed]` with no want/got (exactly the "only default-verbosity FAIL captured" gap). With the linker path set, `TestCUDAForwardMatchesRef` **PASSES** (cosine=1.00000000, argmax-exact).

**Ground truth on L4/sm_89** (`fak-cuda-build-l4`, CUDA 12.9, go1.26.5, libfakcuda.a from the same repro tree):

FAIL:
- `TestCUDAQ8MatMulApproxMatchesRef` — `Q8 argmax-exact failed: ref=265 cudaHost=33 cudaKernel=33` (`cuda_quant_test.go:121`)
- `TestCUDAQ4KMatMulApproxMatchesRef` — `Q4_K argmax-exact failed: ref=6 cudaHost=255 cudaKernel=255` (`cuda_quant_test.go:305`)
- `TestCUDAMatMulNonAlignedShapes` — `out=255 in=256: cosine 0.851 < 0.9999` (`cuda_test.go:244`; the known #972)

PASS (localizes the bug):
- `TestCUDAForwardMatchesRef` (flash-attn forward, cosine 1.0), `TestCUDAGraphDecodeParity` (argmax-exact)
- **both** `TestCUDAQ8BatchedMatMulApproxMatchesRef` **and** `TestCUDAQ4KBatchedMatMulApproxMatchesRef`
- F16 / AWQ (#926) / GPTQ (#3030) MatMul

**Three conclusions:**
1. The defect is the **non-batched (GEMV) Q8_0 / Q4_K quantized MatMul** — the **batched** variants pass, so the general GEMM is fine; it's the single-vector decode path.
2. `cudaHost == cudaKernel` on both (33==33, 255==255) — the device kernel **agrees** with the host dequant and both disagree with the true CPU `ref`, so this is the **shared non-batched dequant / weight-layout**, *not* a device-kernel-only numeric bug. The passing batched path is the reference to mirror.
3. **Rules out** the CUDA-graph-decode-replay and flash-attn hypotheses from the issue body — `TestCUDAGraphDecodeParity` and `TestCUDAForwardMatchesRef` both pass. (The "GPU util 0% during decode" observation is likely just the tiny 0.5B GEMV finishing between samples, not a skipped-compute signal.)

**Serve linkage is now exact:** qwen2.5-0.5b is **Q8_0** and decode is single-token → GEMV → hits `TestCUDAQ8MatMulApproxMatchesRef`'s broken path → wrong argmax → all-`!`.

**So next-step #1 is genuinely done** (failing op named); the fix target is the non-batched Q8_0/Q4_K GEMV dequant/layout in `internal/compute` (`cuda_quant_test.go` cases 121/305). Honest gap unchanged: build tree is a non-git working copy, exact SHA unpinned. `fak-cuda-build-l4` is being reaped again now.

**Comment by anthony-chaudhary** [comment]
**Refinement to the "batched is the reference" line above — there is only ONE kernel.**

Reading the dispatch: both `cudaBackend.MatMul` (P=1) and `cudaBackend.BatchedMatMul` (P>1) call the **same** device kernel —
- Q8_0: `C.fcuda_q8_matmul_f32(…, C.int(P), …)` at `internal/compute/cuda.go:990` (P=1) and `:1030` (P>1)
- Q4_K: `C.fcuda_q4k_matmul_f32(…, C.int(P))` at `:997` (P=1) and `:1037` (P>1)

kernels defined in `internal/compute/cuda_kernels.cu`. So there is **no separate "correct batched kernel" to mirror** — the passing `…Batched…` test and the failing non-batched test exercise the *same* code at different `P`. The failure therefore is one of:
1. a **P=1 (single-row / decode-GEMV) boundary** bug in `fcuda_q8_matmul_f32` / `fcuda_q4k_matmul_f32` (grid/block degeneracy, an off-by-one in the row loop, or the on-device activation-quantization when there's one activation row), or
2. **data-dependent** — the same small numeric error, where the non-batched test's `dominantRow`/seed config (out=320,in=256, seed 485 / 0x4b4b) flips the argmax and the batched test's config doesn't.

Cheap disambiguation on the build host (before touching kernel math): (a) run the *batched* test with **P=1** and the *non-batched* test's data; (b) run the non-batched test at the batched test's dims/seed. If (a) fails and (b) passes → it's the P=1 boundary; if it tracks the data → it's a precision/argmax-margin issue, not a wiring bug. `cudaHost==cudaKernel` (device output itself is wrong; the CPU ref is right) holds either way.

**Comment by claude-ai-netra** [comment]
## Device witness on H100 (sm_90) — the issue's "next checkable step #1", run live

Ran the native-quantized device-GEMM accuracy tests on a real GPU (the step this issue asked for but that was "blocked for lack of a build host"). Provisioned a **fresh GCP spot H100** for it — on-demand `a3-highgpu-1g` was `STOCKOUT` across all three us-central1 zones (a/b/c); the spot pool had capacity. Box torn down after the run.

**Environment:** NVIDIA H100 80GB HBM3, compute cap **9.0 (sm_90)**, driver 580.159.03, CUDA 12.9, Go 1.26.5. `go test -tags cuda ./internal/compute/`.
(Honest provenance caveat: `gcp_bench` ships the working tree as a tarball with `.git` excluded, so the on-box binary carried no VCS stamp — the code under test is the current `main`-region working tree, not a pinned SHA.)

### Result

| test | verdict | detail |
|---|---|---|
| `TestCUDAQ8MatMulApproxMatchesRef` | ✅ PASS | cosine **0.99999980**, argmax-exact |
| `TestCUDAQ8BatchedMatMulApproxMatchesRef` | ✅ **PASS** | cosine **0.99999969** |
| `TestCUDAQ4KMatMulApproxMatchesRef` | ❌ **FAIL** | argmax-exact: **ref=6, device=255** (host-read AND device-kernel argmax both =255) |
| `TestCUDAQ4KBatchedMatMulApproxMatchesRef` | ✅ PASS | cosine **1.00000000** |
| `TestCUDAQuantVRAMWitness` | ✅ PASS | Q8_0 3.56× / Q4_K 7.11× smaller than f32 |

### What this changes about #3186

1. **The Q8_0 device GEMM is CORRECT on-device now.** `TestCUDAQ8BatchedMatMulApproxMatchesRef` — ledgered here as *failing* on DGX A100 (sm_80, 2026-06-30/07-01) — **passes** on **sm_90** with cosine 0.99999969, as does the single-vector Q8 GEMV. So the Q8 lane's ledgered accuracy failure is resolved on a third arch. By implication, the **all-`!` Q8 serve output can no longer be pinned on a Q8 GEMM-accuracy bug in the current tree** — it either was fixed alongside the GEMM, or its root cause is a higher layer (the decode/CUDA-graph replay path — consistent with the issue's own "0% GPU-util during decode" observation). A live `--backend cuda` serve smoke on the same box is the follow-up to disambiguate.

2. **The Q4_K failure is real, live, and cross-arch — but narrower than "device-GEMM accuracy" writ large.** `TestCUDAQ4KMatMulApproxMatchesRef` still fails on sm_90 (matching the ledgered sm_80 failure): the dequant-fused **GEMV** picks output channel **255** where the cpuref f32 reference picks **6**, and both the host argmax-over-device-output and the device `k_argmax` kernel agree on 255 — so the device output genuinely has its max at the wrong channel.

3. **It's data-dependent, not a P=1 code path.** `TestCUDAQ4KBatchedMatMulApproxMatchesRef` (P=8) passes with cosine **1.0**, and both `MatMul` (P=1) and `BatchedMatMul` (P=8) dispatch to the *same* `fcuda_q4k_matmul_f32` → `k_q4k_gemm`. The two tests only differ in their random seed (`0x4b4b` vs `0x4b4c`) and the aligned-vs-random activation. So the bug is triggered by **specific Q4_K super-block bytes** (the single test's channel-255 block), not by the batch dimension.

### Source cross-check (no obvious code error)

On inspection `k_q4k_gemm`'s dequant-fuse (`w = d·scale·code − dmin·min`, `getScaleMinK4_dev`, the `qi`/`is` sub-block walk) matches `dequantQ4KWeight`/`getScaleMinK4` line-for-line, and `k_q8_gemm` + `uploadQ8` match the cpuref Q8 scheme — consistent with Q8 passing and with the Q4_K failure being a **data-dependent numeric discrepancy** rather than a visible port error.

### Next checkable step

Isolate the channel-255 discrepancy: on a GPU host, in `TestCUDAQ4KMatMulApproxMatchesRef` dump the device-dequant `w_device[255]` vs the reference `w[255]` and the raw 144-byte super-block for channel 255 (`f16 d/dmin`, the 12 packed scale bytes, the 128 code bytes). The `__half` d/dmin read (`__half2float(*(const __half*)(blk))`) and `getScaleMinK4_dev`'s `j>=4` branch are the first suspects for a value that inflates one channel. Repro:

```
go test -tags cuda -count=1 -v -run TestCUDAQ4KMatMulApproxMatchesRef ./internal/compute/
# link needs: export LIBRARY_PATH=/usr/local/cuda/lib64  (in addition to -lfakcuda)
```

**Comment by claude-ai-netra** [comment]
Resolved by 2ca497ddda89cebbd0cdd2388dfe96f8a6d3cbc3. The close is independently witnessed by dos commit-audit: OK / diff-witnessed, and origin/main contains the DCO-signed two-file commit. Root cause: the global CUDA upload cache treated mutable rank-one token activations as immutable pointer-keyed weights, so a reused host address returned the prior token's device buffer and made decode deterministic and degenerate. Before: TestCUDAForwardMatchesRef failed at prefill cosine -0.999700 below 0.9990. After: the new TestCUDAUploadRankOneCopiesCurrentContents passes; TestCUDAForwardMatchesRef passes 6 prompt plus 8 greedy tokens with exact argmax and final cosine 0.99998528; CUDA graph, Q8, Q4_K, and six non-aligned MatMul cases also pass. A live real-model Qwen2.5-0.5B CUDA run generated 16 greedy tokens, matched the f32 reference exactly, and had 12 distinct token IDs (non-degenerate). The optional HTTP wrapper smoke could not compile because of the unrelated committed-tip build break now tracked in #4248; the requested model/backend generation path is nevertheless live-witnessed. The issue's secondary fail-closed readiness-probe finding is preserved separately as #4247. Reopen if the reported CUDA decode collapse recurs.
